adaptive/src/main/java/com/google/accompanist/adaptive/TwoPane.kt

```diff
@@ -0,0 +1,411 @@
+                    val height = constraints.maxHeight - splitBottom
+                    constraints.copy(minHeight = height, maxHeight = height)
+                }
+            val firstPlaceable = firstMeasurable.measure(constraints.constrain(firstConstraints))
+            val secondPlaceable = secondMeasurable.measure(constraints.constrain(secondConstraints))
+
+            firstPlaceable.placeRelative(0, 0)
+            val detailOffsetX =
+                if (isHorizontalSplit) {
+                    constraints.maxWidth - secondPlaceable.width
+                } else {
+                    0
+                }
+            val detailOffsetY =
+                if (isHorizontalSplit) {
+                    0
+                } else {
+                    constraints.maxHeight - secondPlaceable.height
+                }
+            secondPlaceable.placeRelative(detailOffsetX, detailOffsetY)
+        }
+    }
+}
+
+/**
+ * A strategy for configuring the [TwoPane] component, that is responsible for the meta-data
+ * corresponding to the arrangement of the two panes of the layout.
+ */
+@Stable
+fun interface TwoPaneStrategy {
+    /**
+     * Calculates the split result in local bounds.
+     */
+    fun calculateSplitResult(
+        density: Density,
+        layoutDirection: LayoutDirection,
+        layoutCoordinates: LayoutCoordinates
+    ): SplitResult
+}
+
+/**
+ * Returns the specification for where to place a split in [TwoPane] as a result of
+ * [TwoPaneStrategy.calculateSplitResult]
+ */
+class SplitResult(
+
+    /**
+     * Whether the split is vertical or horizontal
+     */
+    val isHorizontalSplit: Boolean,
+
+    /**
+     * The area that is nether a `start` pane or an `end` pane, but a separation between those two.
+     * In case width or height is 0 - it means that the area itself is a 0 width/height, but the
+     * place within the layout is still defined.
+     */
+    val splitArea: Rect,
+)
+
+/**
+ * Returns a [TwoPaneStrategy] that will place the slots vertically or horizontally if there is a
+ * horizontal or vertical fold respectively.
+ *
+ * If there is no fold, then the [fallbackStrategy] will be used instead.
+ */
+public fun TwoPaneStrategy(
+    fallbackStrategy: TwoPaneStrategy,
+    windowGeometry: WindowGeometry,
+): TwoPaneStrategy = HorizontalTwoPaneStrategy(
+    fallbackStrategy = VerticalTwoPaneStrategy(
+        fallbackStrategy = fallbackStrategy,
+        windowGeometry = windowGeometry,
+    ),
+    windowGeometry = windowGeometry
+)
+
+/**
+ * Returns a [TwoPaneStrategy] that will place the slots horizontally if there is a vertical fold.
+ *
+ * If there is no horizontal fold, then the [fallbackStrategy] will be used instead.
+ */
+public fun HorizontalTwoPaneStrategy(
+    fallbackStrategy: TwoPaneStrategy,
+    windowGeometry: WindowGeometry,
+): TwoPaneStrategy = TwoPaneStrategy { density, layoutDirection, layoutCoordinates ->
+    val verticalFold = windowGeometry.displayFeatures.find {
```

Comment:
Yep that's true and makes sense, I was trying to keep them together as similarly-scoped pieces of information. I think this relates to a larger `CompositionLocal` question:

If we keep them separate, then that's asking for more window-layout information to be passed around everywhere in the app through different layers. But we keep explicitness, and it's easy to replace in previews and tests.

If we instead provide them in a `CompositionLocal`, then we lose the explicitness,and previews and testing is more difficult (needing to override a `CompositionLocal` instead of passing it manually). But, we avoid having to pass them through everywhere, and we can also make it easier to have the strategies be fold-aware by default.

Reply:
Regarding the split, I think splitting them we don't necessarily ask for more movements, but it might feel like because we operate with two entities instead of one. With that in mind, we might notice that we don't propagate WSC where it is not needed, so in that matter we actually have better-SRP-ed API that asks only for what it needs.

CompositionLocal, as you mentioned, adds some implicit-y and we can always fall back to this option if we want to. Do you think it would make sense to try without it? If developers want, they can do it themselves easy-peasy. Also if you want to read this composition local, you will have to deal with the @composables everywhere here in the API, which I'm not sure is a good idea for such a flexible API surface as we have right now.

Could you try to split them and we can see how it goes? Maybe then it would warrant for a composition local for the display features or the object that contains them? And we can keep it as a parameter on the two pane so that we can default to the composition local, and allow to override for previews and tests.

Reply:
Updated to split up `WindowGeometry`, and depend just on the list of `DisplayFeature`s!